kolibri/core/assets/src/views/sync/SelectDeviceModalGroup/useDevices.js

```diff
@@ -129,51 +143,83 @@ function useDevicesWithFilter(apiParams, filterFunction) {
 }
 
 /**
- * Filters devices to those that also have the specified channel
- * @param channelId
- * @return {{
- *  devices: Ref<NetworkLocation[]>, hasFetched: Ref<bool>, isFetching: Ref<bool>,
- *  filterFailed: Ref<bool>, forceFetch: (function(): Promise<void>)
- * }}
+ * Produces a memoized function that returns a Promise resolving with a boolean for filtering
+ * devices, and automatically clears memoized result if it fails, unless it is a 404
+ *
+ * @param {function(NetworkLocation): Promise<boolean>} filterFunction
+ * @return {(function(NetworkLocation): Promise<boolean>)}
  */
-export function useDevicesWithChannel(channelId = '') {
-  // If channelId is not provided, then we are at top-level import workflow and do not
-  // disable any devices unless it is unavailable
-  const filterDevices =
-    channelId === '' ? () => Promise.resolve(true) : channelIsAvailableAtDevice.bind({}, channelId);
+function useAsyncDeviceFilter(filterFunction) {
+  const memoized = useMemoize(filterFunction, { getKey: device => device.id });
 
-  return useDevicesWithFilter({}, filterDevices);
+  return async function deviceFilter(device) {
+    try {
+      return await memoized(device);
+    } catch (e) {
+      // If not 404 clear cache to try again on next poll
+      if (e?.response?.status !== 404) {
+        memoized.cache.delete(device.id);
+        throw e;
+      }
+
+      return false;
+    }
+  };
 }
 
 /**
- * Filters devices to those that also have the specified facility
- * @param facilityId
- * @return {{
- *  devices: Ref<NetworkLocation[]>, hasFetched: Ref<bool>, isFetching: Ref<bool>,
- *  filterFailed: Ref<bool>, forceFetch: (function(): Promise<void>)
- * }}
+ * Produces a function that resolves with a boolean for a device that has the specified facility
+ * @param {string|null} [id]
+ * @param {bool|null} [learner_can_sign_up]
+ * @return {function(NetworkLocation): Promise<boolean>}
+ */
+export function useDeviceFacilityFilter({ id = null, learner_can_sign_up = null }) {
+  const filters = {};
+
+  // If `id` is an empty string, we don't want to filter by that
+  if (id) {
+    filters.id = id;
+  }
+
+  if (learner_can_sign_up !== null) {
+    filters.learner_can_sign_up = learner_can_sign_up;
+  }
+
+  if (Object.keys(filters).length === 0) {
+    return () => Promise.resolve(true);
+  }
+
+  return useAsyncDeviceFilter(function deviceFacilityFilter(device) {
```

Comment:
Does the parameter passed to `useAsyncDeviceFilter` have to be named or could it be an anonymous function?

Reply:
Mostly asking because when I see something named, especially as specifically as `deviceFacilityFilter` I kind of expect it to be something that I need to know about elsewhere or later but it seems like this could just as well be called `filterFn` or nothing at all -- although I could totally be missing something elsewhere?

Reply:
It doesn't have to be name. Although, there is a benefit when reading error stack traces, because anonymous functions would of course be anonymous and indistinguishable from other anonymous.